### PR TITLE
Upstream (and rename) sortish sampler

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR moves the logic of the "sortish sampler" from examples/seq2seq utils to `trainer_pt_utils` to make this behavior available for all types of training in the main `Trainer`. It also fixes a bug in the previous implementation of the distributed sortish sampler that did not synchronize the random generator used for the shuffling (thus the data returned on the two processes joined together was not a permutation of the whole dataset).

The sortish sampler logic is to group items of the training datasets that have similar lengths together to minimize padding while retaining a bit of randomness. It does some sorting for this, but that's not the main feature, and it's unclear for anyone reading it what it might do, so the argument name was badly chose in my opinion. I chose to name it `group_by_length` when introducing it in `TrainingArguments` (while keeping the old `sortish_sampler` argument in `Seq2SeqTrainingArguments` for backward compatibility, for now).

The actual samplers are given by the two introduced classes `LengthGroupedSampler` and `DistributedLengthGroupedSampler`. They are both tested in the test, and in particular, the distributed one has a test it is using the same random generator for the bit of randomness.

Renaming the old `sortish_sampler` arg is just done in the tests of seq2seq examples for now, it will be done more generally when the seq2seq finetuning script is rewritten to use `datasets`.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -448,14 +449,32 @@ def _get_train_sampler(self) -> Optional[torch.utils.data.sampler.Sampler]:
...
-                if self.args.local_rank == -1
-                else DistributedSampler(self.train_dataset)
-            )
+            num_processes = 1
+            process_index = 0
+
+        # Build the sampler.
+        if self.args.group_by_length:
+            if num_processes <= 1:
+                return LengthGroupedSampler(self.train_dataset, self.args.train_batch_size)
```
Comment (patrickvonplaten) [inline src/transformers/trainer.py:467]: I like the naming!

File: src/transformers/trainer_pt_utils.py [hunk 2]
```diff
@@ -390,3 +391,136 @@ def __call__(self, model_output, labels):
...
+    """
+    Return a list of indices so that each slice of :obj:`batch_size` consecutive indices correspond to elements of
+    similar lengths. To do this, the indices are:
+
+    - randomly permuted
+    - grouped in mega-batches of size :obj:`mega_batch_mult * batch_size`
+    - sorted by length in each mega-batch
+
+    The result is the concatenation of all mega-batches, with the batch of :obj:`batch_size` containing the element of
+    maximum length placed first, so that an OOM happens sooner rather than later.
```
Comment (patrickvonplaten) [inline src/transformers/trainer_pt_utils.py:406]: Ahh, now I get why we put the largest element first - very nice!

**Review by patrickvonplaten — APPROVED** [review]
Great work, very clean!

File: src/transformers/trainer_pt_utils.py (hunk 2 above)
Comment (LysandreJik) [reply src/transformers/trainer_pt_utils.py:406]: This is extremely cool - wouldn't have thought of it!

File: src/transformers/trainer_pt_utils.py [hunk 3]
```diff
@@ -390,3 +391,136 @@ def __call__(self, model_output, labels):
...
+            mega_batch_mult = 1
+
+    # We need to use torch for the random part as a distributed sampler will set the random seed for torch.
+    indices = torch.randperm(len(lengths), generator=generator)
+    megabatch_size = mega_batch_mult * batch_size
+    megabatches = [indices[i : i + megabatch_size].tolist() for i in range(0, len(lengths), megabatch_size)]
+    megabatches = [list(sorted(megabatch, key=lambda i: lengths[i], reverse=True)) for megabatch in megabatches]
+
+    # The rest is to get the biggest batch first.
+    # Since each meagbatch is sorted by descending length, the longest element is the first
```
Comment (LysandreJik) [inline src/transformers/trainer_pt_utils.py:422]:
```suggestion
    # Since each megabatch is sorted by descending length, the longest element is the first
```

**Review by LysandreJik — APPROVED** [review]
This is very, very cool! Thanks for working on it @sgugger!
